File: ShiftMaster.Employee.Service/Application/Services/EmployeeService.cs
using Microsoft.Extensions.Logging;
using ShiftMaster.Employee.Service.Application.DTOs;
using ShiftMaster.Employee.Service.Application.Interfaces;
using ShiftMaster.Employee.Service.Domain.Entities;
using ShiftMaster.Employee.Service.Domain.Enums;
using ShiftMaster.Employee.Service.Domain.Interfaces;
using ShiftMaster.Employee.Service.Infrastructure.Messaging;

namespace ShiftMaster.Employee.Service.Application.Services;

public class EmployeeService : IEmployeeService
{
    private readonly IEmployeeRepository _repository;
    private readonly IOrganisationFilterRepository _orgRepository;
    private readonly IEventPublisher _eventPublisher;
    private readonly ILogger<EmployeeService> _logger;

    public EmployeeService(
        IEmployeeRepository repository,
        IOrganisationFilterRepository orgRepository,
        IEventPublisher eventPublisher,
        ILogger<EmployeeService> logger)
    {
        _repository = repository;
        _orgRepository = orgRepository;
        _eventPublisher = eventPublisher;
        _logger = logger;
    }

    public async Task<EmployeeProfileDto?> GetMeAsync(Guid userId, CancellationToken ct = default)
    {
        var employee = await _repository.GetByUserIdAsync(userId.ToString(), ct);
        return MapToProfile(employee);
    }

    public async Task<EmployeeProfileDto?> GetByIdAsync(Guid id, CancellationToken ct = default)
    {
        var employee = await _repository.GetByIdAsync(id, ct);
        return MapToProfile(employee);
    }

    public async Task<EmployeeListResponse> GetListAsync(EmployeeListRequest request, CancellationToken ct = default)
    {
        var (items, total) = await _repository.GetListAsync(
            request.Search, request.Pole, request.Cellule, request.Department, request.Role,
            request.Page, request.PageSize, ct);
        var dtos = items.Select(e => new EmployeeListItemDto(
            e.Id.ToString(),
            $"{e.FirstName} {e.LastName}".Trim(),
            e.Email,
            e.JobTitle,
            e.Department,
            e.CellId,
            e.Role,
            e.Status.ToString(),
            e.ContractType.ToString(),
            e.Seniority.ToString(),
            e.EquityScore,
            e.LeaveBalance,
            e.Skills.Select(s => s.Name).ToList()
        )).ToList();
        return new EmployeeListResponse(dtos, total, request.Page, request.PageSize);
    }

    public async Task<OrganisationFilterDto> GetOrganisationFiltersAsync(CancellationToken ct = default)
    {
        var f = await _orgRepository.GetDistinctFiltersAsync(ct);
        return new OrganisationFilterDto(f.Poles, f.Cellules, f.Departments);
    }

    public async Task<EmployeeProfileDto?> CreateAsync(CreateEmployeeRequest request, CancellationToken ct = default)
    {
        if (await _repository.GetByEmailAsync(request.Email, ct) != null)
            return null;

        var id = Guid.NewGuid();
        var employee = new Employee
        {
            Id = id,
            UserId = request.UserId,
            FirstName = request.FirstName,
            LastName = request.LastName,
            Email = request.Email,
            JobTitle = request.JobTitle,
            Department = request.Department,
            CellId = request.CellId,
            Role = request.Role,
            Status = EmployeeStatus.Active,
            ContractType = Enum.TryParse<ContractType>(request.ContractType, out var ctEnum) ? ctEnum : ContractType.CDI,
            Seniority = Enum.TryParse<Seniority>(request.Seniority, out var seEnum) ? seEnum : Seniority.Mid,
            EquityScore = 80,
            LeaveBalance = request.LeaveBalance,
            CreatedAt = DateTime.UtcNow,
            Skills = (request.Skills ?? []).Select(s => new EmployeeSkill { Id = Guid.NewGuid(), EmployeeId = id, Name = s, Level = 2 }).ToList(),
            AvailabilitySlots = CreateDefaultAvailability(id)
        };
        await _repository.AddAsync(employee, ct);
        return MapToProfile(await _repository.GetByIdAsync(id, ct));
    }

    public async Task<EmployeeProfileDto?> UpdateAsync(Guid id, UpdateEmployeeRequest request, CancellationToken ct = default)
    {
        var employee = await _repository.GetByIdAsync(id, ct);
        if (employee == null) return null;

        if (request.FirstName != null) employee.FirstName = request.FirstName;
        if (request.LastName != null) employee.LastName = request.LastName;
        if (request.Email != null) employee.Email = request.Email;
        if (request.JobTitle != null) employee.JobTitle = request.JobTitle;
        if (request.Department != null) employee.Department = request.Department;
        if (request.CellId != null) employee.CellId = request.CellId;
        if (request.Role != null) employee.Role = request.Role;
        if (request.Status != null && Enum.TryParse<EmployeeStatus>(request.Status, out var st)) employee.Status = st;
        if (request.ContractType != null && Enum.TryParse<ContractType>(request.ContractType, out var c)) employee.ContractType = c;
        if (request.Seniority != null && Enum.TryParse<Seniority>(request.Seniority, out var s)) employee.Seniority = s;
        if (request.LeaveBalance.HasValue) employee.LeaveBalance = request.LeaveBalance.Value;
        if (request.Skills != null)
        {
            employee.Skills.Clear();
            foreach (var name in request.Skills)
                employee.Skills.Add(new EmployeeSkill { Id = Guid.NewGuid(), EmployeeId = id, Name = name, Level = 2 });
        }
        employee.UpdatedAt = DateTime.UtcNow;
        await _repository.SaveChangesAsync(ct);
        return MapToProfile(await _repository.GetByIdAsync(id, ct));
    }

    public Task<bool> DeleteAsync(Guid id, CancellationToken ct = default) => _repository.DeleteAsync(id, ct);

    private static List<AvailabilitySlot> CreateDefaultAvailability(Guid employeeId)
    {
        var slots = new List<AvailabilitySlot>();
        for (var d = DayOfWeek.Monday; d <= DayOfWeek.Friday; d++)
            slots.Add(new AvailabilitySlot { Id = Guid.NewGuid(), EmployeeId = employeeId, DayOfWeek = d, StartTime = new TimeOnly(8, 0), EndTime = new TimeOnly(18, 0), IsAvailable = true });
        return slots;
    }

    private static EmployeeProfileDto? MapToProfile(Domain.Entities.Employee? employee)
    {
        if (employee == null) return null;

        var name = $"{employee.FirstName} {employee.LastName}".Trim();
        return new EmployeeProfileDto(
            employee.Id.ToString(),
            employee.UserId,
            employee.FirstName,
            employee.LastName,
            name,
            employee.Email,
            employee.JobTitle,
            employee.Department,
            employee.CellId,
            employee.AvatarUrl,
            employee.Role,
            employee.Status.ToString(),
            employee.ContractType.ToString(),
            employee.Seniority.ToString(),
            employee.EquityScore,
            employee.LeaveBalance,
            employee.Skills.Select(s => s.Name).ToList(),
            employee.AvailabilitySlots.Select(a => new AvailabilitySlotDto(
                a.DayOfWeek, a.StartTime, a.EndTime, a.IsAvailable)).ToList()
        );
    }
}

File: src/app/core/services/employee.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { API_BASE } from './api.config';

export interface EmployeeListItem {
  id: string;
  name: string;
  email: string;
  jobTitle: string;
  department: string;
  cellId: string;
  role: string;
  status: string;
  contractType: string;
  seniority: string;
  equityScore: number;
  leaveBalance: number;
  skills: string[];
}

export interface EmployeeListResponse {
  items: EmployeeListItem[];
  totalCount: number;
  page: number;
  pageSize: number;
}

export interface EmployeeProfile {
  id: string;
  userId: string;
  firstName: string;
  lastName: string;
  name: string;
  email: string;
  jobTitle: string;
  department: string;
  cellId: string;
  avatarUrl?: string;
  role: string;
  status: string;
  contractType: string;
  seniority: string;
  equityScore: number;
  leaveBalance: number;
  skills: string[];
  availabilitySlots?: { dayOfWeek: number; startTime: string; endTime: string; isAvailable: boolean }[];
}

export interface OrganisationFilters {
  poles: string[];
  cellules: string[];
  departments: string[];
}

export interface CreateEmployeeRequest {
  userId: string;
  firstName: string;
  lastName: string;
  email: string;
  jobTitle: string;
  department: string;
  cellId: string;
  role: string;
  contractType: string;
  seniority: string;
  leaveBalance: number;
  skills?: string[];
}

export interface UpdateEmployeeRequest {
  firstName?: string;
  lastName?: string;
  email?: string;
  jobTitle?: string;
  department?: string;
  cellId?: string;
  role?: string;
  status?: string;
  contractType?: string;
  seniority?: string;
  leaveBalance?: number;
  skills?: string[];
}

export interface EmployeeListParams {
  search?: string;
  pole?: string;
  cellule?: string;
  department?: string;
  role?: string;
  page?: number;
  pageSize?: number;
}

@Injectable({ providedIn: 'root' })
export class EmployeeService {
  private readonly base = `${API_BASE}/employees`;

  constructor(private http: HttpClient) {}

  getList(params: EmployeeListParams = {}): Observable<EmployeeListResponse> {
    let httpParams = new HttpParams();
    if (params.search) httpParams = httpParams.set('search', params.search);
    if (params.pole) httpParams = httpParams.set('pole', params.pole);
    if (params.cellule) httpParams = httpParams.set('cellule', params.cellule);
    if (params.department) httpParams = httpParams.set('department', params.department);
    if (params.role) httpParams = httpParams.set('role', params.role);
    if (params.page != null) httpParams = httpParams.set('page', params.page.toString());
    if (params.pageSize != null) httpParams = httpParams.set('pageSize', params.pageSize.toString());
    return this.http.get<EmployeeListResponse>(this.base, { params: httpParams });
  }

  getOrganisationFilters(): Observable<OrganisationFilters> {
    return this.http.get<OrganisationFilters>(`${this.base}/organisation-filters`);
  }

  getById(id: string): Observable<EmployeeProfile> {
    return this.http.get<EmployeeProfile>(`${this.base}/${id}`);
  }

  getMe(): Observable<EmployeeProfile> {
    return this.http.get<EmployeeProfile>(`${this.base}/me`);
  }

  getMeEquity(cellId?: string): Observable<{ score: number; [key: string]: unknown }> {
    const params = cellId ? { cellId } : {};
    return this.http.get<{ score: number; [key: string]: unknown }>(`${this.base}/me/equity`, { params });
  }

  getEquity(id: string, cellId?: string): Observable<{ score: number; [key: string]: unknown }> {
    const params = cellId ? { cellId } : {};
    return this.http.get<{ score: number; [key: string]: unknown }>(`${this.base}/${id}/equity`, { params });
  }

  create(request: CreateEmployeeRequest): Observable<EmployeeProfile> {
    return this.http.post<EmployeeProfile>(this.base, request);
  }

  update(id: string, request: UpdateEmployeeRequest): Observable<EmployeeProfile> {
    return this.http.put<EmployeeProfile>(`${this.base}/${id}`, request);
  }

  delete(id: string): Observable<void> {
    return this.http.delete<void>(`${this.base}/${id}`);
  }
}
